lib/mastodon/snowflake.rb

```diff
@@ -4,25 +4,11 @@ module Mastodon::Snowflake
   DEFAULT_REGEX = /timestamp_id\('(?<seq_prefix>\w+)'/
 
   class Callbacks
-    def self.around_create(record)
+    def self.before_create(record)
       now = Time.now.utc
 
-      if record.created_at.nil? || record.created_at >= now || record.created_at == record.updated_at
-        yield
-      else
-        record.id = Mastodon::Snowflake.id_at(record.created_at)
-        tries     = 0
-
-        begin
-          yield
-        rescue ActiveRecord::RecordNotUnique
-          raise if tries > 100
-
-          tries     += 1
-          record.id += rand(100)
-
-          retry
-        end
+      if record.created_at.present? && record.created_at < now && record.created_at != record.updated_at
+        record.id = Mastodon::Snowflake.id_at(record.class.table_name, record.created_at)
```

Comment:
I gather you're doing this because `ActivityPub::Activity::Create#process_status` operates in a transaction, and so the RecordNotUnique could fail the transaction anyway.

This will just fail straight away instead. Is there anything to be gained from doing this, rather than changing line 19 to something like this?

~~~ ruby
raise if (tries > 100) || (ActiveRecord::Base.connection.open_transactions > 0)
~~~

That has the same effect inside a transaction, but not outside a transaction. (I'm not actually sure how frequently this would be called outside a transaction, but it's worth considering, maybe?)

Reply:
Either way, it might be good to handle this outside of the transaction (for example, in `ActivityPub::Activity::Create#process_status`), if we're going to be concerned about the potential for duplicate IDs.

Reply:
> if we're going to be concerned about the potential for duplicate IDs

I'm not so I just removed it. We are not concerning that when we use `timestamp_id` anyway.

Reply:
> if we're going to be concerned about the potential for duplicate IDs

I'm not so I just removed it. We are not concerning that when we use `timestamp_id` anyway.

Reply:
> if we're going to be concerned about the potential for duplicate IDs

I'm not so I just removed it. We are not concerning that when we use `timestamp_id` anyway.

Reply:
> if we're going to be concerned about the potential for duplicate IDs

I'm not so I just removed it. We are not concerning that when we use `timestamp_id` anyway.

Reply:
> if we're going to be concerned about the potential for duplicate IDs

I'm not so I just removed it. We are not concerning that when we use `timestamp_id` anyway.

Reply:
> if we're going to be concerned about the potential for duplicate IDs

I'm not so I just removed it. We are not concerning that when we use `timestamp_id` anyway.

Reply:
> if we're going to be concerned about the potential for duplicate IDs

I'm not so I just removed it. We are not concerning that when we use `timestamp_id` anyway.

Reply:
> if we're going to be concerned about the potential for duplicate IDs

I'm not so I just removed it. We are not concerning that when we use `timestamp_id` anyway.

Reply:
GitHub went mad and posted a same comment so many times. Sorry for messing.